File: SignalServiceKit/src/Messages/DeviceSyncing/PaymentsSyncMessage.m
//
//  Copyright (c) 2021 Open Whisper Systems. All rights reserved.
//

#import "PaymentsSyncMessage.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

@implementation PaymentsSyncMobileCoinOutgoing

- (instancetype)initWithPicoMob:(uint64_t)picoMob
            recipientUuidString:(NSString *)recipientUuidString
                        receipt:(NSData *)receipt
                     blockIndex:(uint64_t)blockIndex
                 blockTimestamp:(uint64_t)blockTimestamp
                 spentKeyImages:(NSArray<NSData *> *)spentKeyImages
               outputPublicKeys:(NSArray<NSData *> *)outputPublicKeys
                    memoMessage:(nullable NSString *)memoMessage
{
    self = [super init];
    if (!self) {
        return nil;
    }

    _picoMob = picoMob;
    _recipientUuidString = recipientUuidString;
    _receipt = receipt;
    _blockIndex = blockIndex;
    _blockTimestamp = blockTimestamp;
    _spentKeyImages = spentKeyImages;
    _outputPublicKeys = outputPublicKeys;
    _memoMessage = memoMessage;

    return self;
}

@end

#pragma mark -

@interface PaymentsSyncMessage ()

@property (nonatomic, readonly, nullable) PaymentsSyncMobileCoinOutgoing *mcOutgoing;

@end

#pragma mark -

@implementation PaymentsSyncMessage

- (instancetype)initWithThread:(TSThread *)thread mcOutgoing:(nullable PaymentsSyncMobileCoinOutgoing *)mcOutgoing
{
    self = [super initWithThread:thread];
    if (!self) {
        return nil;
    }

    _mcOutgoing = mcOutgoing;

    return self;
}

- (nullable instancetype)initWithCoder:(NSCoder *)coder
{
    return [super initWithCoder:coder];
}

- (nullable SSKProtoSyncMessageBuilder *)syncMessageBuilderWithTransaction:(SDSAnyReadTransaction *)transaction
{
    return [self syncMessageBuilderWithMCOutgoing:self.mcOutgoing transaction:transaction];
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/Messages/DeviceSyncing/PaymentsSyncMessage.swift
//
//  Copyright (c) 2021 Open Whisper Systems. All rights reserved.
//

import Foundation

@objc
public extension PaymentsSyncMessage {

    @objc(syncMessageBuilderWithMCOutgoing:transaction:)
    func syncMessageBuilder(mcOutgoing: PaymentsSyncMobileCoinOutgoing?,
                            transaction: SDSAnyReadTransaction) -> SSKProtoSyncMessage.SSKProtoSyncMessageBuilder? {
        // TODO: Support defrags.
        // TODO: Support requests.

        do {
            let paymentBuilder = SSKProtoSyncMessagePayment.builder()

            if let mcOutgoing = mcOutgoing {
                paymentBuilder.setOutgoing(try outgoingBuilder(mcOutgoing: mcOutgoing))
            }

            let builder = SSKProtoSyncMessage.builder()
            builder.setPayment(try paymentBuilder.build())
            return builder
        } catch {
            owsFailDebug("Error: \(error)")
            return nil
        }
    }

    private func outgoingBuilder(mcOutgoing: PaymentsSyncMobileCoinOutgoing) throws -> SSKProtoSyncMessagePaymentOutgoing {
        let outgoingBuilder = SSKProtoSyncMessagePaymentOutgoing.builder()

        let mobileCoinBuilder = SSKProtoSyncMessagePaymentOutgoingMobileCoin.builder(recipientUuid: mcOutgoing.recipientUuidString,
                                                                                     picoMob: mcOutgoing.picoMob,
                                                                                     receipt: mcOutgoing.receipt,
                                                                                     ledgerBlockIndex: mcOutgoing.blockIndex)
        if mcOutgoing.blockTimestamp > 0 {
            mobileCoinBuilder.setLedgerBlockTimestamp(mcOutgoing.blockTimestamp)
        }
        if !mcOutgoing.spentKeyImages.isEmpty {
            mobileCoinBuilder.setSpentKeyImage(mcOutgoing.spentKeyImages)
        }
        if !mcOutgoing.outputPublicKeys.isEmpty {
            mobileCoinBuilder.setOutputPublicKey(mcOutgoing.outputPublicKeys)
        }
        if let memoMessage = mcOutgoing.memoMessage {
            mobileCoinBuilder.setMemoMessage(memoMessage)
        }

        outgoingBuilder.setMobileCoin(try mobileCoinBuilder.build())

        return try outgoingBuilder.build()
    }
}
